core/domain/classifier_services.py

```diff
@@ -279,27 +305,37 @@ def _update_classifier_training_jobs_status(job_ids, status):
     """
     classifier_training_job_models = (
         classifier_models.ClassifierTrainingJobModel.get_multi(job_ids))
+    strict_classifier_training_job_models: List[
```

Comment:
This seems odd. The typing should not make it to the "variable names" level. 

Could we not just type classifier_training_job_models as a list? I don't really follow what's going on here and I think developers will find it hard to understand the difference between the two models and what they should do to each.

Reply:
Problem: The type of `classifier_training_job_models` is `Sequence[Optional[ClassifierTrainingJobModel]]` and the expected type for argument of `classifier_models.ClassifierTrainingJobModel.update_timestamps_multi()` is `List[ClassifierTrainingJobModel]`. So, if we narrow down the type of `classifier_training_job_models` from Sequence to List ( type after narrowing down `List[Optional[ClassifierTrainingJobModel]]` ) then still this type is not acceptable by `classifier_models.ClassifierTrainingJobModel.update_timestamps_multi()` because it require `List[ClassifierTrainingJobModel]` but we are providing `List[Optional[ClassifierTrainingJobModel]]`

Reply:
There are 2 solutions that we can apply here:
1. Use a `cast` to convert the type of `classifier_training_job_models` from `Sequence[Optional[ClassifierTrainingJobModel]]` to `List[ClassifierTrainingJobModel]`. But I'll not prefer this.
2. Introduce a new variable ( like I did, introduced `strict_classifier_training_job_models` ) and append all the models one by one after doing an assertion ( like in my case I appended all the models in line `326`, and the assertion is done in line `314` ). After doing this, no `None` value is present in `strict_classifier_training_job_models` and MyPy will consider its type as `List[ClassifierTrainingJobModel]`. So now the type matches the argument's type of `classifier_models.ClassifierTrainingJobModel.update_timestamps_multi()` and MyPy will not throw error.

Reply:
I have discussed this issue with Aasif and we found that this issue is encountered already but the solution is not implemented yet!!! you can see it here https://github.com/python/mypy/issues/4573

Reply:
It seems to have been "closed as completed". Maybe check with them what's going on?

Also conceptually it seems to me that the difference is "list of models with possible Nones" and "list of models that are guaranteed to be models". It's fine to have two separate variables for those two things if that's what you actually want to distinguish -- in that case, it's more a question of the logical concept rather than something that's just for the type-checks. If we actually would need to distinguish them even in the case where we didn't have type-checking then I'm fine with different variables but these need to have better names to actually explain what they are.

Reply:
Ok, I'll write them an email about asking the situation of the issue and their future planning regarding this issue, right?

For names:
shall we use `classifier_training_job_models_with_none` for `list of models with possible Nones`
and 
`classifier_training_job_models` for `list of models that are guaranteed to be models` ???

Reply:
I think you haven't really answered my question. Do we actually need to treat these as separate logical concepts throughout the code, if we weren't doing typing?

Reply:
> I think you haven't really answered my question. Do we actually need to treat these as separate logical concepts throughout the code, if we weren't doing typing?

Yes, we should consider it as a separate concept. Because there are cases in the codebase where we are not checking the presence of None in List ( or Sequence ) and just passing the values. 

Just want to mention that most of the cases are encountered after introducing typing!!

Reply:
That's good! OK then, let's have two different variables for two different concepts, named appropriately.

Your names sound fine to me. They're better than the previous names, since "strict' doesn't really convey "not none" in general.

Reply:
Done! Name is changed!!